lib/senders/email.js

```diff
@@ -253,97 +267,227 @@ module.exports = function (log, config) {
+          }
 
-    var emailConfig = {
-      sender: this.sender,
-      from: this.sender,
-      to: message.email,
-      subject: localized.subject,
-      text: localized.text,
-      html: localized.html,
-      xMailer: false,
-      headers
-    }
+          log.info({
+            email: emailAddresses[0],
+            op: 'mailer.send',
+            template,
+            headers: Object.keys(headers).join(',')
+          })
+
+          const emailConfig = {
+            sender: this.sender,
+            from: this.sender,
+            to: emailAddresses[0],
+            subject: localized.subject,
+            text: localized.text,
+            html: localized.html,
+            xMailer: false,
+            headers
+          }
 
-    // Utilize nodemailer's cc ability to send to multiple addresses
-    // Ref: https://nodemailer.com/message/
-    if (message.ccEmails) {
-      emailConfig.cc = message.ccEmails
-    }
+          if (emailAddresses.length > 1) {
+            emailConfig.cc = emailAddresses.slice(1)
+          }
 
-    var d = P.defer()
-    mailer.sendMail(
-      emailConfig,
-      function (err, status) {
-        if (err) {
-          log.error(
-            {
+          const d = P.defer()
+          mailer.sendMail(emailConfig, (err, status) => {
+            if (err) {
+              log.error({
+                op: 'mailer.send.1',
+                err: err && err.message,
+                status: status && status.message,
+                id: status && status.messageId,
+                to: emailConfig && emailConfig.to
+              })
+
+              return d.reject(err)
+            }
+
+            log.info({
               op: 'mailer.send.1',
-              err: err && err.message,
               status: status && status.message,
               id: status && status.messageId,
               to: emailConfig && emailConfig.to
+            })
+
+            emailUtils.logEmailEventSent(log, message)
+
+            return d.resolve(status)
+          })
+
+          return d.promise
+        }))
+      })
+  }
+
+  // Based on the to and cc email addresses of a message, return an array of
+  // `Service` objects that control how email traffic will be routed.
+  //
+  // It will attempt to read live config data from Redis and live config takes
+  // precedence over local static config. If no config is found at all, email
+  // will be routed locally via the auth server.
+  //
+  // Live config looks like this (every property is optional):
+  //
+  // {
+  //   sendgrid: {
+  //     percentage: 100,
+  //     regex: "^.+@example\.com$"
+  //   },
+  //   ses: {
+  //     percentage: 10,
+  //     regex: ".*",
+  //   }
+  // }
+  //
+  // Where a percentage and a regex are both present, an email address must
+  // satisfy both criteria to count as a match. Where an email address matches
+  // both sendgrid and ses, sendgrid wins.
```

Comment:
Does this exposition make sense? Is it clear and unambiguous about how the different configs interact?

(and actually, are those interactions themselves sensible choices?)

Reply:
I had to read through the comment and the code a couple of times for everything to make sense in my head.

* It'd be useful to make explicit that both fields are optional. It's implied already but I had to read through the code to verify.
* If both fields are present, it can be thought of "N percentage of emails that match `regex` will be sent via that service" If only a regex, then 100% of matching emails are sent via that service, and if only a percentage, then N % of all emails sent via that service.

Reply:
> It'd be useful to make explicit that both fields are optional.

Just checking, as it does say `every property is optional` earlier in the comment, you mean more explicit than that?

Reply:
Perhaps I need new glasses.